crates/linker/src/macho.rs

```diff
@@ -601,453 +630,292 @@ fn gen_macho_le(
-    // This is what the elf version currently does.
-    // Theoretically it is not needed (if we update all virtual addresses in the binary),
-    // but I definitely ran into problems with elf when not adding this extra buffering.
-
-    // Copy header and shift everything to enable more program sections.
-    let added_byte_count = ((2 * segment_cmd_size) + (2 * section_size) - total_cmd_size) as u64;
-    md.added_byte_count = added_byte_count
-        // add some alignment padding
-        + (MIN_SECTION_ALIGNMENT as u64 - added_byte_count % MIN_SECTION_ALIGNMENT as u64);
-
-    md.exec_len = exec_data.len() as u64 + md.added_byte_count;
+    let available_size =
+        md.start_of_first_section - (size_of_cmds + size_of_header - total_cmd_size);
 
-    let mut out_mmap = open_mmap_mut(out_filename, md.exec_len as usize);
-    let end_of_cmds = size_of_cmds + mem::size_of_val(exec_header);
-
-    // "Delete" the dylib load command - by copying all the bytes before it
-    // and all the bytes after it, while skipping over its bytes.
-    // It has a dynamic-length string at the end that we also need to delete,
-    // in addition to the header.
-    out_mmap[..macho_load_so_offset].copy_from_slice(&exec_data[..macho_load_so_offset]);
-
-    out_mmap[macho_load_so_offset..end_of_cmds - total_cmd_size]
-        .copy_from_slice(&exec_data[macho_load_so_offset + total_cmd_size..end_of_cmds]);
+    if available_size < required_size {
+        internal_error!("Not enough free space between end of load commands and start of first section\nConsider recompiling the host with -headerpad <size> linker flag");
+    }
 
-    // Copy the rest of the file, leaving a gap for the surgical linking to add our 2 commands
-    // (which happens after preprocessing), and some zero padding at the end for alignemnt.
-    // (It seems to cause bugs if that padding isn't there!)
-    let rest_of_data = &exec_data[end_of_cmds..];
-    let start_index = end_of_cmds + md.added_byte_count as usize;
+    md.exec_len = exec_data.len() as u64;
 
-    out_mmap[start_index..start_index + rest_of_data.len()].copy_from_slice(rest_of_data);
+    let mut out_mmap = open_mmap_mut(out_filename, md.exec_len as usize);
+    let end_of_cmds = size_of_cmds + mem::size_of_val(exec_header);
 
-    let out_header = load_struct_inplace_mut::<macho::MachHeader64<LE>>(&mut out_mmap, 0);
+    // Copy load commands over making space for Roc-specific commands
+    let start_of_roc_commands = {
+        let mut out_offset = 0;
+        let mut in_offset = 0;
+        let mut size = md.linkedit_segment;
+
+        out_mmap[out_offset..out_offset + size]
+            .copy_from_slice(&exec_data[in_offset..in_offset + size]);
+        out_offset += size;
+        in_offset += size;
+
+        // Make space for Roc segment load commands
+        let start_of_roc_commands = out_offset;
+        out_offset += required_size;
+        md.linkedit_segment += required_size;
```

Comment:
Why does this get updated? I thought that we were fitting in the gap between the load commands and the link edit section, so it wouldn't be moved.

Reply:
Oh, I missed the comment in the description that the link edit section is actually at the end of the binary and we are inserting before it. Some reason, I thought that it was right after the load commands.

Reply:
Yes, correct. The layout of an executable image (incl a dylib) is as follows:

```
__PAGEZERO (optional for dylibs)
__TEXT
__DATA_CONST
__DATA
__LINKEDIT
```

and `__LINKEDIT` always has to come last, so we save the offset of the segment load command describing it so that we can quickly shift it by a required amount when copying over the data from the app in the surgical linking step.